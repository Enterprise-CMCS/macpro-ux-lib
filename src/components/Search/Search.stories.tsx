import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Search } from "./Search";

export default {
  title: "COMPONENTS/Search",
  component: Search,
  argTypes: {
    variation: {
      description: "Renders the style of the search component.",
    },
    disabled: {
      description: "Determines whether or not a button is enabled.",
    },
    value: {
      description: "The searched value within the input.",
    },
    placeholder: {
      description: "The text placeholder for this component.",
    },
    onSearch: {
      description: "Handles its behavior when the button is clicked.",
      control: false,
    },
  },
  args: {
    placeholder: "Placeholder here",
    onSearch: (e) => console.log("test", e.target.value),
  },
} as ComponentMeta<typeof Search>;

const Template: ComponentStory<typeof Search> = ({ ...rest }) => (
  <Search {...rest} />
);

export const PrimarySearch = Template.bind({});
PrimarySearch.args = {
  variation: "default",
};

export const BigSearch = Template.bind({});
BigSearch.args = {
  variation: "big",
};

export const SmallSearch = Template.bind({});
SmallSearch.args = {
  variation: "small",
};
