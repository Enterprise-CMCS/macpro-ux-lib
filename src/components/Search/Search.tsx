import React, { EventHandler, FormEventHandler, useState } from "react";
type IntrinsicElements = JSX.IntrinsicElements["input"];

interface Props extends IntrinsicElements {
  variation?: SearchVariation;
  onSearch: EventHandler<any>;
  onInput?: FormEventHandler<HTMLInputElement>;
  disabled?: boolean;
  initialValue?: string;
  placeholder?: string;
  labelText?: string;
}
type SearchVariation = "default" | "big" | "small";

/**
 * Search Component
 * @param {string}            variation             Renders the style of the search component.
 * @param {EventHandler}      onSearch              Handles its behavior when the search button is clicked or the user presses enter in the input. Returns value of the input. Returns value of the input.
 * @param {EventHandler}      onInput               Handles its behavior when the user is typing.
 * @param {boolean}           [disabled]            Determines whether or not a button is enabled.
 * @param {string}            [initialValue]        If provided, used as value of search input.
 * @param {string}            [placeholder]         The text placeholder for this component.
 * @param {string}            [labelText]           Label text for screen reader.
 */

export const Search: React.FC<Props> = ({
  variation = "default",
  disabled = false,
  onSearch,
  initialValue = "",
  labelText = "search",
  onInput,
  ...rest
}) => {
  const [inputValue, setInputValue] = useState<string>(initialValue);

  return (
    <section aria-label="Search component">
      <form
        className={`usa-search role usa-search--${variation}`}
        role="search"
      >
        <label className="usa-sr-only" htmlFor="search-field">
          {labelText}
        </label>
        <input
          className="usa-input"
          id="search-field"
          type="search"
          name="search"
          value={inputValue}
          disabled={disabled}
          onInput={onInput}
          onChange={(e) => setInputValue(e.target.value)}
          onKeyDown={(e) => {
            if (e.key === "Enter") {
              e.preventDefault();
              onSearch(inputValue);
            }
          }}
          {...rest}
        />
        <button
          className="usa-button padding-y-0 padding-x-2 radius-left-0"
          type="button"
          disabled={disabled}
          onClick={() => onSearch(inputValue)}
        >
          {variation !== "small" && (
            <span className="usa-search__submit-text">Search </span>
          )}
          <img
            src="/assets/img/usa-icons-bg/search--white.svg"
            className="usa-search__submit-icon"
            alt="Search"
          />
        </button>
      </form>
    </section>
  );
};
