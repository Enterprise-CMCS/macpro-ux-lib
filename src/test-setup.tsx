import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
