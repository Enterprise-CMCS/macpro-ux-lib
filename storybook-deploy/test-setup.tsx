import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
