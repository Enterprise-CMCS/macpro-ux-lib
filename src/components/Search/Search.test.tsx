import React from "react";
import { Search } from "./Search";
import { fireEvent, screen, render } from "../../test-setup";

describe("Tests for the search component.", () => {
  const mockChangeFn = jest.fn();

  it("Should render and fire a callback", () => {
    render(
      <Search
        data-testid="search"
        variation="default"
        placeholder="search"
        onSearch={mockChangeFn}
      />
    );
    const searchComp = screen.getByTestId("search");

    fireEvent.keyDown(searchComp, {
      key: "Enter",
      code: "Enter",
      charCode: 13,
    });
    expect(searchComp).toBeInTheDocument();
    expect(searchComp).toHaveAttribute("placeholder", "search");
    expect(mockChangeFn).toHaveBeenCalled();
  });

  it("Should render a search bar with an icon", () => {
    render(
      <Search
        variation="small"
        placeholder="Test Search"
        data-testid="search"
        onSearch={mockChangeFn}
        onInput={mockChangeFn}
      />
    );
    const searchComp = screen.getByTestId("search");

    fireEvent.change(searchComp, { target: { value: "Good Day" } });

    expect(searchComp).toBeInTheDocument();
    expect(mockChangeFn).toHaveBeenCalled();
    expect(searchComp).toHaveAttribute("placeholder", "Test Search");
    expect(searchComp).toHaveAttribute("id", "search-field-en-small");
  });

  it("Should render a search bar with a value and the button simulates a search event", () => {
    render(
      <Search
        variation="big"
        placeholder="Test Search"
        data-testid="search"
        onSearch={mockChangeFn}
        initialValue="new value"
        labelText="Search Label"
      />
    );
    const searchComp = screen.getByTestId("search");
    const buttonComp = screen.getByLabelText("Search Label");

    expect(searchComp).toHaveAttribute("value", "new value");
    expect(buttonComp).toBeInTheDocument();

    fireEvent.click(buttonComp);
    expect(buttonComp).toBeInTheDocument();
    expect(mockChangeFn).toHaveBeenCalled();
  });

  describe("compontent snapshots", () => {
    it("primary big disabled", () => {
      const { container } = render(
        <Search
          onSearch={mockChangeFn}
          data-testid="search"
          variation="default"
          labelText="Search Click Here"
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("secondary big search", () => {
      const { container } = render(
        <Search
          onSearch={mockChangeFn}
          initialValue="initial"
          data-testid="search"
          variation="big"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
