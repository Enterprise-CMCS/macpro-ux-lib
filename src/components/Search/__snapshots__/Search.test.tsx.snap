// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the search component. compontent snapshots big search 1`] = `
<div>
  <section
    aria-label="Search component"
  >
    <form
      class="usa-search role usa-search--big"
      role="search"
    >
      <label
        class="usa-sr-only"
        for="search-field"
      >
        search
      </label>
      <input
        class="usa-input"
        data-testid="search"
        id="search-field"
        name="search"
        type="search"
        value="initial"
      />
      <button
        aria-label="search"
        class="usa-button padding-y-0 padding-x-2 radius-left-0"
        type="button"
      >
        <span
          class="usa-search__submit-text"
        >
          Search 
        </span>
        <img
          alt="Search"
          class="usa-search__submit-icon"
          src="[object Object]"
        />
      </button>
    </form>
  </section>
</div>
`;

exports[`Tests for the search component. compontent snapshots default search 1`] = `
<div>
  <section
    aria-label="Search component"
  >
    <form
      class="usa-search role usa-search--default"
      role="search"
    >
      <label
        class="usa-sr-only"
        for="search-field"
      >
        Search Click Here
      </label>
      <input
        class="usa-input"
        data-testid="search"
        id="search-field"
        name="search"
        type="search"
        value=""
      />
      <button
        aria-label="search"
        class="usa-button padding-y-0 padding-x-2 radius-left-0"
        type="button"
      >
        <span
          class="usa-search__submit-text"
        >
          Search 
        </span>
        <img
          alt="Search"
          class="usa-search__submit-icon"
          src="[object Object]"
        />
      </button>
    </form>
  </section>
</div>
`;

exports[`Tests for the search component. compontent snapshots small disabled search 1`] = `
<div>
  <section
    aria-label="Search component"
  >
    <form
      class="usa-search role usa-search--small"
      role="search"
    >
      <label
        class="usa-sr-only"
        for="search-field"
      >
        Click Me
      </label>
      <input
        class="usa-input"
        data-testid="search"
        disabled=""
        id="search-field"
        name="search"
        placeholder="placeholder"
        type="search"
        value=""
      />
      <button
        aria-label="search"
        class="usa-button padding-y-0 padding-x-2 radius-left-0"
        disabled=""
        type="button"
      >
        <img
          alt="Search"
          class="usa-search__submit-icon"
          src="[object Object]"
        />
      </button>
    </form>
  </section>
</div>
`;
